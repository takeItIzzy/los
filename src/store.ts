import createSubscribe, { StateBucket, Subscribe } from './utils/createSubscribe';

/**
 * Collecting state to store when useLosState or init/useLosInit is executing.
 * useLosValue triggers component re-rendering when state updated.
 */

export interface StoreItem {
  hasInit: boolean;
  value: any;
  reducer?: LosReducer<any, any>;
  stateBucket: StateBucket;
  subscribe: Subscribe;
}
export const store: Map<Atom<any, any>, StoreItem> = new Map();

export type LosAction<T> = { type: T; [key: string]: any };
export type LosReducer<S, A> = (state: S, action: LosAction<A>) => S;
export class Atom<T, A = void> {
  constructor(value?: T, reducer?: LosReducer<T, A>) {
    this.value = value;
    this.reducer = reducer;
  }
  value?: T;
  reducer?: LosReducer<T, A>;
}

export const atom = <T, A = void>(
  config: {
    defaultValue?: T;
    reducer?: LosReducer<T, A>;
  } = {}
): Atom<T, A> => {
  const { defaultValue, reducer } = config;
  const atomItem = new Atom(defaultValue, reducer);

  const stateBucket: StateBucket = new Set();
  store.set(atomItem, {
    hasInit: false,
    value: atomItem.value,
    reducer: atomItem.reducer,
    stateBucket,
    subscribe: createSubscribe(stateBucket),
  });

  return atomItem;
};

interface ComputedGetMethods {
  get: <T, A>(atom: Atom<T, A>) => T;
}
interface ComputedSetMethods extends ComputedGetMethods {
  set: <T, A>(atom: Atom<T, A>, value: T) => void;
}
export class Computed<Derive> {
  constructor(
    getter: ({ get }: ComputedGetMethods) => Derive,
    setter?: (config: ComputedSetMethods, newValue: Derive) => void
  ) {
    this.getter = getter;
    this.setter = setter;
    this.originAtoms = new Set<Atom<any, any>>();
    this.stateProvider = this.stateProvider.bind(this);
  }
  getter: (config: ComputedGetMethods) => Derive;
  setter?: (config: ComputedSetMethods, newValue: Derive) => void;
  originAtoms: Set<Atom<any, any>>;
  stateProvider<T, A>(atom: Atom<T, A>): T {
    this.originAtoms.add(atom);
    return store.get(atom)!.value;
  }
  get value() {
    return this.getter({ get: this.stateProvider });
  }
}

type ComputedConfig<Derive> = {
  get: (methods: ComputedGetMethods) => Derive;
  set?: <NewValue extends Derive>(methods: ComputedSetMethods, value: NewValue) => void;
};
export const computed = <Derive>(config: ComputedConfig<Derive>): Computed<Derive> => {
  const { get, set } = config;

  return new Computed(get, set);
};
