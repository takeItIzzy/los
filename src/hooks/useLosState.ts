import * as React from 'react';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
import { Atom, store } from '../store';
import mergeStoreItem from '../utils/mergeStoreItem';
import pushSubscribe from '../utils/pushSubscribe';

export const useLosValue = <T, A = void>(atom: Atom<T, A>): T => {
  const value = useSyncExternalStore(
    store.get(atom)!.subscribe,
    React.useCallback(() => store.get(atom)!.value, [atom])
  );

  React.useDebugValue(value);

  return value;
};

type SetStateFunction<T> = (state: T) => T;
export type SetLosState<T> = (state: T | SetStateFunction<T>) => void;
export const useSetLosState = <T, A = void>(atom: Atom<T, A>): SetLosState<T> => {
  return (state: T | SetStateFunction<T>) => {
    mergeStoreItem(atom, {
      // now that we start updating state, we can confirm that the atom has been initialized
      hasInit: true,
      // @ts-ignore
      value: typeof state === 'function' ? state(store.get(atom)!.value) : state,
    });

    pushSubscribe(atom);
  };
};

export const useLosState = <T, A = void>(atom: Atom<T, A>): [T, SetLosState<T>] => {
  return [useLosValue(atom), useSetLosState(atom)];
};
