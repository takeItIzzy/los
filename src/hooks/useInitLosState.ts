import { Atom, store } from '../store';
import { SetLosState, useLosValue, useSetLosState } from './useLosState';
import mergeStoreItem from '../utils/mergeStoreItem';
import pushSubscribe from '../utils/pushSubscribe';
import createSubscribe from '../utils/createSubscribe';

export const initLosState = <T, A = void>(
  state: Atom<T, A>,
  defaultValue: T,
  allowReinitialize?: boolean
): void => {
  if (allowReinitialize) {
    mergeStoreItem(state, {
      hasInit: true,
      value: defaultValue,
      subscribe: createSubscribe(store.get(state)!.stateBucket),
    });
    pushSubscribe(state);
  } else if (!store.get(state)!.hasInit) {
    mergeStoreItem(state, {
      hasInit: true,
      value: defaultValue,
      subscribe: store.get(state)!.subscribe ?? createSubscribe(store.get(state)!.stateBucket),
    });
    pushSubscribe(state);
  }
};

export const useInitLosState = <T, A = void>(
  atom: Atom<T, A>,
  defaultValue: T,
  allowReinitialize?: boolean
): [T, SetLosState<T>] => {
  initLosState(atom, defaultValue, allowReinitialize);

  return [useLosValue(atom), useSetLosState(atom)];
};
