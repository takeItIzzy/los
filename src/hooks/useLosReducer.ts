import { Atom, LosAction, store } from '../store';
import { useLosValue } from './useLosState';
import { initLosState } from './useInitLosState';
import updateStoreItem from '../utils/updateStoreItem';
import { __DEV__ } from '../constants';
import { error } from '../utils/warn';

type LosDispatch<A> = (action: LosAction<A>) => void;

export const losDispatch = <T, A = void>(state: Atom<T, A>): LosDispatch<A> => {
  if (__DEV__ && !(state instanceof Atom)) {
    error('losDispatch: state must be an Atom');
  }
  return (action) => {
    const currentState = store.get(state);
    updateStoreItem(state, {
      // now that we start updating state, we can confirm that the atom has been initialized
      hasInit: true,
      value: currentState!.reducer!(currentState!.value, action),
    });
  };
};

export const useLosDispatch = losDispatch;

export const useLosReducer = <T, A = void>(
  state: Atom<T, A>,
  initStateValue?: T,
  allowReinitialize?: boolean
): [T, LosDispatch<A>] => {
  !!initStateValue && initLosState(state, initStateValue, allowReinitialize);

  return [useLosValue(state), useLosDispatch(state)];
};
