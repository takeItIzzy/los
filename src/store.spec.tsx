import { atom, LosReducer, store } from './store';
import { useLosState } from './hooks/useLosState';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import * as React from 'react';

describe('atom() testing', () => {
  it('should exist current key in store when defaultValue is provided', () => {
    const state = atom({
      defaultValue: 'default',
    });
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: 'default',
      reducer: undefined,
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('should exist current key in store when reducer is provided', () => {
    const reducer: LosReducer<any, any> = (state) => {
      return state;
    };
    const state = atom({
      reducer,
    });
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: undefined,
      reducer: [Function],
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('should exist current key in store when defaultValue and reducer are both provided', () => {
    const reducer: LosReducer<any, any> = (state) => {
      return state;
    };
    const state = atom({
      defaultValue: 'default',
      reducer,
    });
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: 'default',
      reducer: [Function],
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('should exist current key in store when defaultValue and reducer are neither provided', () => {
    const state = atom();
    expect(store.has(state)).toBeTruthy();
    expect(store.get(state)).toMatchSnapshot(`{
      hasInit: false,
      value: undefined,
      reducer: undefined,
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    }`);
  });

  it('if there are no more subscribers and the atom is not persistent, reset the atom to its initial value', async () => {
    const uncachedState = atom({ defaultValue: 0, cached: false });
    const Foo = () => {
      const [state, setState] = useLosState(uncachedState);

      return (
        <>
          <div data-testid="showState">uncached: {state}</div>
          <button data-testid="setStateButton" onClick={() => setState(3)}>
            click me
          </button>
        </>
      );
    };
    const Boo = () => {
      const [state, setState] = useLosState(uncachedState);

      return (
        <>
          <div data-testid="showState2">uncached2: {state}</div>
          <button data-testid="setStateButton2" onClick={() => setState(3)}>
            click me
          </button>
        </>
      );
    };
    const Wrapper = () => {
      const [showFoo, setShowFoo] = React.useState(true);
      const [showBoo, setShowBoo] = React.useState(true);

      return (
        <>
          <button data-testid="showFooButton" onClick={() => setShowFoo((prev) => !prev)}>
            click me
          </button>
          <button data-testid="showBooButton" onClick={() => setShowBoo((prev) => !prev)}>
            click me
          </button>
          {showFoo && <Foo />}
          {showBoo && <Boo />}
        </>
      );
    };

    const user = userEvent.setup();
    render(<Wrapper />);

    await user.click(screen.getByTestId('setStateButton'));
    await user.click(screen.getByTestId('setStateButton2'));

    expect(screen.getByTestId('showState')).toHaveTextContent('uncached: 3');
    expect(screen.getByTestId('showState2')).toHaveTextContent('uncached2: 3');

    await user.click(screen.getByTestId('showFooButton')); // hide Foo
    await user.click(screen.getByTestId('showFooButton')); // show Foo

    // atom won't be reset because there are still subscribers
    expect(screen.getByTestId('showState')).toHaveTextContent('uncached: 3');

    await user.click(screen.getByTestId('showFooButton')); // hide Foo
    await user.click(screen.getByTestId('showBooButton')); // hide Boo
    await user.click(screen.getByTestId('showFooButton')); // show Foo
    await user.click(screen.getByTestId('showBooButton')); // show Boo

    // atom will be reset because all subscribers used to be removed
    expect(screen.getByTestId('showState')).toHaveTextContent('uncached: 0');
    expect(screen.getByTestId('showState2')).toHaveTextContent('uncached2: 0');
  });
});
