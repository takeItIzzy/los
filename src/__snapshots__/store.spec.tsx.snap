// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atom() testing should exist current key in store when defaultValue and reducer are both provided: {
      hasInit: false,
      value: 'default',
      reducer: [Function],
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "cached": true,
  "hasInit": false,
  "reducer": [Function],
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": "default",
}
`;

exports[`atom() testing should exist current key in store when defaultValue and reducer are neither provided: {
      hasInit: false,
      value: undefined,
      reducer: undefined,
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "cached": true,
  "hasInit": false,
  "reducer": undefined,
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": undefined,
}
`;

exports[`atom() testing should exist current key in store when defaultValue is provided: {
      hasInit: false,
      value: 'default',
      reducer: undefined,
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "cached": true,
  "hasInit": false,
  "reducer": undefined,
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": "default",
}
`;

exports[`atom() testing should exist current key in store when reducer is provided: {
      hasInit: false,
      value: undefined,
      reducer: [Function],
      cached: true,
      stateBucket: new Set(),
      subscribe: [Function]
    } 1`] = `
Object {
  "cached": true,
  "hasInit": false,
  "reducer": [Function],
  "stateBucket": Set {},
  "subscribe": [Function],
  "value": undefined,
}
`;
